docs: Exposition on which threads formula evaluation uses. (#7458)
deephaven/deephaven-core: docs/groovy/how-to-guides/performance/formula-threads.md

---
title: Threads for formula evaluation
sidebar_label: Formulas and threads
---

The Deephaven query engine executes as a Java process and makes use of multiple threads to process requests. When the Deephaven engine executes a query, it does so with an ExecutionContext that controls which variables and methods are available to formulas and provides information about the user and query initialization. The query engine can be used as a library, in which case the application developer is responsible for providing a consistent threading model. This guide describes the standard Deephaven server application, as executed via our Docker containers, the Gradle `server-jetty-app:run` task, or as a Core+ worker in the Deephaven Enterprise system.

> Most interaction with the Deephaven engine is via gRPC. When processing a gRPC request, it is initially handled on a web-server thread as part of the Java gRPC library. Depending on the request, it is then dispatched to one of two thread pools for request handling. The Serial executor handles requests using a single thread. Evaluating script commands from a Code Studio and resolving scope tickets execute on the serial queue. Other requests are handled on the concurrent executor, which has four threads by default, but is user-configurable by setting the configuration property `scheduler.poolSize`. The single-threaded executor pool provides a well-defined order for script code execution and resolving variables. Other operations may execute concurrently, with necessary locking handled by the default [Periodic Update Graph](https://deephaven.io/core/groovy/docs/conceptual/periodic-update-graph-configuration/).

Reviewer: ```suggestion
Most interaction with the Deephaven engine is via gRPC. When processing a gRPC request, it is initially handled on a web-server thread as part of the Java gRPC library. Depending on the request, it is then dispatched to one of two thread pools for request handling. The Serial executor handles requests using a single thread. Evaluating script commands from a Code Studio and resolving scope tickets execute on the this serial queue. Other requests are handled on the concurrent executor, which has four threads by default, but is user-configurable by setting the configuration property `scheduler.poolSize`. The single-threaded executor pool provides a well-defined order for script code execution and resolving variables. Other operations may execute concurrently, with necessary locking handled by the default [Periodic Update Graph](https://deephaven.io/core/groovy/docs/conceptual/periodic-update-graph-configuration/).
```

```groovy
x=emptyTable(1).update([Selectable.parse("Thr=java.lang.Thread.currentThread()").withSerial()])
```

```python
from deephaven.table import Selectable
from deephaven import empty_table

x = empty_table(1).update(
    [Selectable.parse("Thr=java.lang.Thread.currentThread()").with_serial()]
)
```

The `withSerial` method indicates to the Deephaven engine that the `Thr` column must be evaluated in order, and therefore it is not multi-threaded. The result is that the table contains `Thread[DeephavenApiServer-Scheduler-Serial-1,5,main]`, indicating that it has executed on the serial executor thread.

To illustrate, we'll remove the `withSerial` method and execute the following query:

```groovy
x=emptyTable(1).update("Thr=java.lang.Thread.currentThread()")
```

```python
from deephaven import empty_table

x = empty_table(1).update(["Thr=java.lang.Thread.currentThread()"])
```

The Deephaven engine may parallelize evaluation, thus resulting in a value of `Thr` of `Thread[OperationInitializationThreadPool-initializationExecutor-3,5,OperationInitializationThreadPool]`, indicating that the formula was evaluated on the operation initialization thread pool.

> Similarly, any time a source table updates, the downstream effects are evaluated by an Update Graph. The default Periodic Update Graph uses a thread pool that has the same number of threads as the machine has processors (the number of threads can be configured by the property `PeriodicUpdateGraph.updateThreads`).

Reviewer: ```suggestion
Similarly, each time a source table updates, the downstream effects are evaluated by an Update Graph. The default Periodic Update Graph uses a thread pool that has the same number of threads as the machine has processors (the number of threads can be configured by the property `PeriodicUpdateGraph.updateThreads`).
```

```groovy
x=timeTable("PT1s").head(2).update("Thr=java.lang.Thread.currentThread()")
```

```python
from deephaven import time_table

x = time_table(1).head(2).update(["Thr=java.lang.Thread.currentThread()"])
```

In this case, the formula is evaluated on one of the update executor threads (e.g., `Thread[PeriodicUpdateGraph-updateExecutor-6,5,PeriodicUpdateGraph-updateExecutors]`).

The `select` and `update` operations behave identically to each other, eagerly computing the result during initialization or in response to a table update.

## `view` and `updateView`

The `view` and `updateView` operations, on the other hand, only compute the result when the result is accessed. This can happen on a variety of threads. For example, when performing another query operation, the results are read from the thread executing that operation. For example:

```groovy order=x,sd
x=emptyTable(1).view("Thr=java.lang.Thread.currentThread()")
sd=x.selectDistinct()
```

```python order=x,sd
from deephaven import empty_table

x = empty_table(1).view(["Thr=java.lang.Thread.currentThread()"])
sd = x.select_distinct()
```

> The value of `Thr` in `y` is `Thread[DeephavenApiServer-Scheduler-Serial-1,5,main]` - the thread that executed the `selectDistinct` operation. However, when viewing the table `x`, the `Thr` column takes on a value like `Thread[DeephavenApiServer-Scheduler-Concurrent-4,5,main]` because that is the thread that the barrage snapshot operation read the value on. Each time a cell is accessed (e.g., by reloading or scrolling around a table), the value is recomputed potentially on another thread.

Reviewer: `y`? do you mean `sd`?

---
title: Threads for formula evaluation
sidebar_label: Formulas and threads
---

The Deephaven query engine executes as a Java process and makes use of multiple threads to process requests. When the Deephaven engine executes a query, it does so with an [`ExecutionContext`](https://docs.deephaven.io/core/javadoc/io/deephaven/engine/context/ExecutionContext.html) that controls which variables and methods are available to formulas and provides information about the user and query initialization. The query engine can be used as a library, in which case the application developer is responsible for providing a consistent threading model. This guide describes the standard Deephaven server application, as executed via our Docker containers, the Gradle `server-jetty-app:run` task, or as a Core+ worker in the Deephaven Enterprise system.

> Most interaction with the Deephaven engine is via gRPC. When processing a gRPC request, it is initially handled on a web-server thread as part of the Java gRPC library. Depending on the request, it is then dispatched to one of two thread pools for request handling. The serial executor handles requests using a single thread. Evaluating script commands from a Code Studio and resolving scope tickets execute on the serial queue. Other requests are handled on the concurrent executor, which has four threads by default, but is user-configurable by setting the configuration property `scheduler.poolSize`. The single-threaded executor pool provides a well-defined order for script code execution and resolving variables. Other operations may execute concurrently, with necessary locking handled by the default [Periodic Update Graph](https://deephaven.io/core/groovy/docs/conceptual/periodic-update-graph-configuration/).

Reviewer 2: ```suggestion
Most interaction with the Deephaven engine is via gRPC. When processing a gRPC request, it is initially handled on a web-server thread as part of the Java gRPC library. Depending on the request, it is then dispatched to one of two thread pools for request handling. The serial executor handles requests using a single thread. Evaluating script commands from a Code Studio and resolving scope tickets execute on the serial queue. Other requests are handled on the concurrent executor, which has four threads by default, but is user-configurable by setting the configuration property `scheduler.poolSize`. The single-threaded executor pool provides a well-defined order for script code execution and resolving variables. Other operations may execute concurrently, with necessary locking handled by the default [Periodic Update Graph](../../conceptual/periodic-update-graph-configuration.md).
```

> A query operation begins on one of these thread pools, but evaluation may move to another thread depending on the operation. Consider the following snippet executed from a code studio:

Reviewer 2: ```suggestion
A query operation begins on one of these thread pools, but evaluation may move to another thread depending on the operation. Consider the following snippet executed from a Code Studio:
```

```groovy
thread_name=emptyTable(1).update([Selectable.parse("Thr=java.lang.Thread.currentThread().getName()").withSerial()])
```

```python
from deephaven.table import Selectable
from deephaven import empty_table

thread_name = empty_table(1).update(
    [Selectable.parse("Thr=java.lang.Thread.currentThread().getName()").with_serial()]
)
```

> The `withSerial` method indicates to the Deephaven engine that the `Thr` column must be evaluated in order, and therefore it is not multi-threaded. The result is that the table contains `DeephavenApiServer-Scheduler-Serial-1`, indicating that it has executed on the serial executor thread.

Author: I don't like this change.
